### PR TITLE
[Go] Ensure iterated and emitted types are registered.

### LINKED ISSUE #23889: [Bug][Go SDK]: register.Iter* and register.Emit* don't register parameter types.
### What happened?

At present, the regsiter.Function* and register.Dofn* functions register types for their parameters, but the iterables and emitters do not. This leaves gaps in type registration, forcing extra calls to beam.RegisterType, when the package should be handling it.

The solution is to add the RegisterType calls to the methods like we do for DoFn, Combine and Function.

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-go

### PR DESCRIPTION
**Pull request by lostluck**
Fixes #23889

Resolves an issue with a pattern for transform package authors where they receive a given type from users, where that received type wouldn't be registered, leading to potential issues at authoring and testing time.

Transforms registered with `register` won't have types in their emitters or iterators registered due to complexities WRT generics. Instead, also ensure those types are registered with the Iter and Emitter* methods.

Also moves the iterator and emitter tests to co-aligned files, to make them easier to find.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/go/pkg/beam/register/iter.go [hunk 1]
```diff
@@ -113,7 +123,9 @@ func Iter1[T any]() {
 	registerFunc := func(s exec.ReStream) exec.ReusableInput {
 		return &iter1[T]{s: s}
 	}
-	exec.RegisterInput(reflect.TypeOf(i).Elem(), registerFunc)
+	itT := reflect.TypeOf(i).Elem()
```
Comment (capthiron) [inline sdks/go/pkg/beam/register/iter.go:126]: Hey @lostluck 
I just started a test run with your changes applied with the bigtable io connector (#23411).
Unfortunately I still ran into the following error: `reflect.Set: value of type struct { RowKey string; Ops []struct { Family string; Column string; Ts int64; Value []uint8 }; GroupKey string } is not assignable to type bigtableio.Mutation`.
I played around with this implementation but wasn't able to make it work. My knowledge on the reflect package is somehow limited though.

**Comment by lostluck** [comment]
That's very strange. That's what the additional registrations in this PR
should be handling.

That error means it didn't have the type registered and had to
synthetically generate it. In simple cases this satisfies Go's
Assignability rules but not so kuch for nested structs.

Is there more of a panic trace where that error comes from?

**Comment by lostluck** [comment]
Ah, I'm neglecting to strip the pointer from the type. Turns out "derived"
types like pointers and slices "don't exist" with how this registry was
built. Easy enough to fix (and test). Thanks for catching it!

**Comment by lostluck** [comment]
RFAL

**Review by riteshghorse — APPROVED** [review]
The changes looks good to me, thanks!

**Comment by riteshghorse** [comment]
Run Go PostCommit

**Pull request merged**